add gptqmodel support (#2247)
huggingface/peft: docs/source/developer_guides/quantization.md

config = LoraConfig(target_modules="all-linear", ...)
```

## GPTQ quantization

> You can learn more about gptq based `[2, 3, 4, 8]` bits quantization at [GPTQModel](https://github.com/ModelCloud/GPTQModel) and [HF GPTQ Doc](https://github.com/huggingface/transformers/blob/main/docs/source/en/quantization/gptq.md). PEFT post-quant training can use both [GPTQModel](https://github.com/ModelCloud/GPTQModel) or [AutoGPTQ](https://github.com/autogptq/autogptq) libraries but we recommend GPTQModel as AutoGPTQ will be deprecated in a future release. 

Reviewer: ```suggestion
You can learn more about gptq based `[2, 3, 4, 8]` bits quantization at [GPTQModel](https://github.com/ModelCloud/GPTQModel) and the Transformers [GPTQ](https://huggingface.co/docs/transformers/quantization/gptq) doc. Post-quant training, PEFT can use both [GPTQModel](https://github.com/ModelCloud/GPTQModel) or [AutoGPTQ](https://github.com/autogptq/autogptq) libraries, but we recommend GPTQModel because AutoGPTQ will be deprecated in a future release. 
```

```bash
# gptqmodel install
pip install gptqmodel --no-build-isolation
```

```py
from transformers import AutoModelForCausalLM, AutoTokenizer, GPTQConfig

model_id = "facebook/opt-125m"
tokenizer = AutoTokenizer.from_pretrained(model_id)

gptq_config = GPTQConfig(bits=4, group_size=128, dataset="wikitext2", tokenizer=tokenizer)

> quantized_model = AutoModelForCausalLM.from_pretrained(model_id, device_map="auto", quantization_config=gptq_config)

Reviewer 2: When trying to run this locally with 2 CUDA devices, I encountered a CUDA error after 50% progress:

```
File ~/work/forks/transformers/src/transformers/models/opt/modeling_opt.py:559, in OPTDecoderLayer.forward(self, hidden_states, attention_mask, layer_head_mask, past_key_value, output_attentions, use_cache, position_ids)
    556 hidden_states = self.fc2(hidden_states)
    557 hidden_states = nn.functional.dropout(hidden_states, p=self.dropout, training=self.training)
--> 559 hidden_states = (residual + hidden_states).view(hidden_states_shape)
    561 # 350m applies layer norm AFTER attention
    562 if not self.do_layer_norm_before:

RuntimeError: CUDA error: an illegal memory access was encountered
```

Is this a known problem? Using 1 CUDA device or setting `CUDA_LAUNCH_BLOCKING=1` resolves the error.

I suspect that the error occurs at the "switch" from GPU 0 to GPU 1, since that's exactly after half the layers when using `device_map="auto"`.

Reviewer 3: > When trying to run this locally with 2 CUDA devices, I encountered a CUDA error after 50% progress:

Which test is this exactly? We can try to replicate.

Reviewer 2: Agreed that opt is very outdated at this point, and we mainly use it since it's small, but at least for PEFT it hasn't caused any problems yet.

I ran the code above using `meta-llama/Llama-3.2-1B` and again got an error after 50%:

```
File ~/anaconda3/envs/peft/lib/python3.11/site-packages/torch/nn/modules/linear.py:125, in Linear.forward(self, input)
    124 def forward(self, input: Tensor) -> Tensor:
--> 125     return F.linear(input, self.weight, self.bias)

RuntimeError: CUDA error: CUBLAS_STATUS_EXECUTION_FAILED when calling `cublasGemmEx( handle, opa, opb, m, n, k, &falpha, a, CUDA_R_16F, lda, b, CUDA_R_16F, ldb, &fbeta, c, CUDA_R_16F, ldc, CUDA_R_32F, CUBLAS_GEMM_DEFAULT_TENSOR_OP)`
```

Thus it's unlikely to be related to the model architecture. `CUDA_LAUNCH_BLOCKING=1` again was enough to resolve the issue.

# save quantized model
quantized_model.save_pretrained("./opt-125m-gptq")
tokenizer.save_pretrained("./opt-125m-gptq")
```

> Once quantized, you can post-train gptq models using normal PEFT apis.

Reviewer: ```suggestion
Once quantized, you can post-train GPTQ models with PEFT APIs.
```
